File: java/746.min-cost-climbing-stairs.java
/*
 * @lc app=leetcode id=746 lang=java
 *
 * [746] Min Cost Climbing Stairs
 */

// @lc code=start
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        int[] stairCosts = new int[n + 1];

        int currCost;

        for(int i = 2; i <= n; i++) {
            stairCosts[i] = Math.min(stairCosts[i-1] + cost[i-1], stairCosts[i-2] + cost[i-2]);
        }

        return stairCosts[n];
    }
}
// @lc code=end

File: cpp/746.min-cost-climbing-stairs.cpp
/*
 * @lc app=leetcode id=746 lang=cpp
 *
 * [746] Min Cost Climbing Stairs
 */

// @lc code=start
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        vector<int> dp(3, INT_MAX);
        dp[0] = 0;
        dp[1] = 0;

        for(int i = 0; i < cost.size(); i++) {
            dp[(i + 1) % 3] = std::min(dp[(i + 1) % 3], dp[i % 3] + cost[i]);
            dp[(i + 2) % 3] = std::min(dp[(i + 2) % 3], dp[i % 3] + cost[i]);
            dp[i % 3] = INT_MAX;
        }

        return dp[cost.size() % 3];
    }
};
// @lc code=end

